codalab/model/bundle_model.py

```diff
@@ -699,6 +699,78 @@ def get_bundle_uuids(self, conditions, max_results):
 
         return self._execute_query(query)
 
+    def get_memoized_bundles(self, user_id, command, dependencies):
+        '''
+        Get a list of bundle UUIDs that match with input command and dependencies in the order they were created.
+        :param user_id: a string that specifies the current user id.
+        :param command: a string that defines the command that is used to search for memoized bundles in the database.
+        :param dependencies: a string in the form of '[{"child_path": key1, "parent_uuid": uuid1},
+                                                       {"child_path": key2, "parent_uuid": uuid2}]'
+                             to search for memoized bundles in the database.
+        :return: a list of matched UUIDs in the order they were created.
+        '''
+        # Decode json formatted dependencies string to a list of key value pairs
+        dependencies = json.loads(dependencies)
+        # When there is no dependency to be matched, the target memozied bundle
+        # should only exist in the bundle table but not in the bundle_dependency table.
+        if len(dependencies) == 0:
+            query = (
+                select([cl_bundle.c.uuid])
+                .select_from(cl_bundle)
+                .where(
+                    and_(
+                        cl_bundle.c.command == command,
+                        cl_bundle.c.owner_id == user_id,
+                        cl_bundle.c.uuid.notin_(
+                            select([cl_bundle_dependency.c.child_uuid]).select_from(
```

Comment:
How efficient is this query?

Reply:
Benchmarked in both my local machine and staging environment, it appears that in the staging environment, UNION on JOINs runs 5X faster than subqueries. My local machine shows the opposite result but I think we should choose the method that's the best for the production environment. I have updated the code with UNION on JOINs.

Reply:
> How efficient is this query?

I did several rounds of testing on two platforms (MacOS and Staging environment).  
| Tests | 2 AGG in 1 query | 2 AGG in 2 queries | UNIONs on JOINs | 
| ---- | ---- | ---- | ---- | 
| 1| 0.0578651428222656| 0.06195163726806640 | 0.060456275939941406 | 
| 2| 0.5580108165740967| 0.07441258430480957 | 0.06279182434082031 | 
| 3| 0.6198058128356934| 0.06659507751464844 | 0.11051750183105469 | 
| 4| 0.6372816562652588| 0.07648229598999023 | 0.10235214233398438 | 
| 5| 0.6172430515289307| 0.06659984588623047 | 0.06983256340026855 | 
| 6| 0.5474967956542969| 0.06769084930419922 | 0.10195350646972656 | 
| 7| 0.3473646640777588| 0.06359648704528809 | 0.09937453269958496 | 
| 8| 0.3533980846405029| 0.06468772888183594 | 0.10466766357421875 | 
| 9| 0.3261570930480957| 0.05560111999511719 | 0.10058188438415527 | 

Note that *AGG* stands for aggregation.